### PR TITLE
BUG: adding Buhring's analytic continuation method to special.hyp2f1; #8054

### LINKED ISSUE #8054: Gauss Hypergeometric function (hyp2f1) diverges in two small regions of the complex plane
There are two narrow wedges in the complex plane where gauss's hypergeometric function (hyp2f1) gives implausibly large values when the function there is well defined.  In polar coordinates these wedges are approximately given by:

0.9 < r < 1.0
pi/3 < theta < pi/2

0.9 < r < 1.0
-pi/2 < theta < -pi/3

When I looked at the source code (the subroutine HYGFZ in scipy/special/specfun/specfun.f), I noticed that in the half annulus defined by 0.9 < |z| < 1.0 and real(z)>0, HYGFZ attempts to use the "1-z" transformation (see [eq 15.8.4](http://dlmf.nist.gov/15.8)) to evaluate the hypergeometric function. However the series associated with this transformation only converges when |1-z|<1. In the wedge regions given above, |1-z|>1 and so the method fails. 

### Reproducing code example:
```
>>> import numpy as np
>>> from scipy.special import hyp2f1
>>> theta=np.linspace(np.pi/3.0,np.pi/2.0,10,dtype=np.float)
>>> z=0.95*np.exp(1.j*theta)
>>> hyp2f1(1.54,1.67,5.21,z)
array([ -2.14630605e+03 +6.71599145e+03j,
        -2.31991069e+14 +1.69689842e+14j,
        -2.50252397e+24 -3.16737713e+23j,
        -2.97640356e+33 -4.56037250e+33j,
         1.21082099e+42 -3.00909950e+42j,
         5.82402044e+50 -8.54809272e+49j,
         2.01184470e+58 +2.92349095e+58j,
        -4.16032387e+65 +6.43973917e+65j,
        -6.17050365e+72 -1.47233313e+72j,  -1.20089983e+78 -2.12896064e+79j])
```
In contrast, the first value in the z array should be (according to mpmath):
```
>>> import mpmath
>>> mpmath.hyp2f1(1.54,1.67,5.21,z[0])
mpc(real='0.9894407614789662', imag='0.52327754379886682')
```

### Scipy/Numpy/Python version information:
scipy, numpy, sys versions
```
>>> print sc.__version__,np.__version__,sys.version_info
1.0.0.dev0+b579459 1.13.1 sys.version_info(major=2, minor=7, micro=12, releaselevel='final', serial=0)
```

### LINKED ISSUE #8083: special.hyp2f1 returns the wrong values when c-a-b is an integer (for 0.9<|z|<1.0)
The hypergeometric function hyp2f1(a,b,c,z) gets the wrong values when c-a-b=integer and z is in the region of the complex plane given by:

0.9<|z|<1.0 and real(z)>0

The source code for hyp2f1 can definitely handle the case of integer differences between the a,b,c parameters, but in this case the code doesn't correctly calculate the integer corresponding to c-a-b.
Specifically, this bug is caused by a couple of lines in specfun.f in which the integer is determined by the expression
```
M=INT(C-A-B)
```
For some values of a,b,c this might be 3.9999 which gets cast to the integer *3* instead of *4*.  Using a conversion of the form 
```
M=INT(C-A-B+EPS*DSIGN(1.0D0,C-A-B))
```
where EPS is a small value like 1e-8 or 1e-15 should fix this issue (this is at least what specfun.f does elsewhere in HYGFZ when determining an integer from real values). 

### Reproducing code example:
```
>>> import numpy as np
>>> from scipy.special import hyp2f1 
>>> import mpmath
>>> a=1.45
>>> b=0.6
>>> c=6.05
>>> c-a-b
3.9999999999999996
>>> z=0.9*np.exp(1.j*(-0.2*np.pi))
>>> hyp2f1(a,b,c,z)
(5.670408102465732-0.76807186486958579j)
>>> mpmath.hyp2f1(a,b,c,z)
mpc(real='1.1053555735636691', imag='-0.11840327506397474')
```
In the above test case, HYGFZ is using *3* instead of *4* for c-a-b which is why it gets the wrong answer as compared with mpmath.

### Scipy/Numpy/Python version information:
```
>>> print scipy.__version__,np.__version__,sys.version_info
1.0.0.dev0+b579459 1.13.1 sys.version_info(major=2, minor=7, micro=12, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by FormerPhysicist**
## Overview
This pull request adds a new method of calculating the hypergeometric function hyp2f1 in the region where the standard transformation formulas for hyp2f1 fail. This method is based on the analytic continuation series given in _W. Buhring, An Analytic Continuation of the Hypergeometric Series, SIAM J. Math. Anal., Vol. 18, No. 3, May 1987_.

This new method is applied to the region of the complex plane given by:
0.9 <= |z| <= 1.1 and Re(z) >= 0.0 and |z-1.0| >= 0.75

## Summary of what methods are used in each region of the complex plane

|z| > 1.1  ............................................................................  1/z transformation
|z| <= 1.1 and Re(z) < 0.0  ................................................  z/(z-1) transformation
|z| < 0.9 and Re(z) >= 0.0 .................................................  defining Taylor series for hyp2f1
0.9 <= |z| <= 1.1 and |z-1.0| < 0.75  .................................. 1-z transformation
0.9 <= |z| <= 1.1 and Re(z) >= 0.0 and |z-1.0| >= 0.75  ....  Buhring's analytic continuation series

Note that the transformations for hyp2f1 discussed above are given [here](http://dlmf.nist.gov/15.8). (They're also listed in _Handbook of Mathematical Functions, M. Abramowitz and I. Stegun_)
## Expanded region that uses 1-z and z/(z-1) transformations 

Besides implementing Buhring's series in the region where hyp2f1 was failing, I also slightly extended the region where the 1-z and z/(z-1) transformations are used so that both transformations are now used (in their appropriate regions) **when |z| <= 1.1** instead of when |z| <= 1.0

My main motivation for this small expansion is the fact that the three methods that would get used in the ring 1.0 < |z| < 1.1 (i.e., Buhring's series, the 1-z and z/(z-1) transformations, see previous summary) all converge much sooner and are much faster than the 1/z transformation.

The only potential (but ultimately fixable) downside that I could find with the expansion of the 1-z region is related to the values we return for hyp2f1 on the branch cut (1, inf).  Specifically, on the branch cut of hyp2f1, the 1-z transformation converges to a different branch of hyp2f1 than the 1/z transformation does. 

From a mathematical point of view this is okay since hyp2f1 is essentially multi-valued on the branch cut but it does mean that if one were to plot the real/imaginary parts of hyp2f1 on the real axis (1,inf), one would see a discontinuity at x=1.1.  To fix this, I've also added some code to the 1-z method so that it converges to the same branch of hyp2f1 that the 1/z transformation converges to (when z is on the branch cut).  This amounts to multiplying some prefactors in the method by a constant complex value.  I can give more details about this as needed.

## Related issues
Closes #8054 
Closes #8083

### CODE REVIEW

**Comment by FormerPhysicist** [comment]
Note that the failure associated with the Travis CI checks is NOT related to my development. It involves some code with 'test_ndimage.py' and/or possibly Numpy deprecation warnings that were added recently...

File: scipy/special/tests/test_hyp2f1.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+from scipy.special._testutils import FuncData, MissingModule, check_version
+from scipy.special import hyp2f1
+
+try:
+    import mpmath
+except ImportError:
+    mpmath = MissingModule('mpmath')
+
+# Fix the seed number
+np.random.seed(1108257)
```
Comment (xoviat) [inline scipy/special/tests/test_hyp2f1.py]: I don't think this is the correct place to put this. It should go on the first line of `test_hyp2f1`.

Comment (FormerPhysicist) [reply scipy/special/tests/test_hyp2f1.py]: @xoviat I moved where the random seed is set

**Comment by brsr** [comment]
Excuse my butting in, but is there anything in particular keeping this request from being pulled? I ran into this bug myself earlier today.

**Comment by rgommers** [comment]
@brsr nothing in particular, it just needs reviewing. @FormerPhysicist apologies that we didn't get to this PR. The issue seems to be that this is quite technical code with only a few of the maintainers having the background (Fortran skills + special functions math) to review this.

I'll mark it for the next release, so we will at least revisit it before then. @person142 any chance you can have a look at this?

**Comment by person142** [comment]
I’ll give it my best shot. Definitely not in the next week though so further pings might be needed...

**Comment by FormerPhysicist** [comment]
@rgommers @person142 No worries about the delay; I keep an eye out over the next view weeks for any new comments/questions and try to respond to any new concerns you guys have.

**Comment by FormerPhysicist** [comment]
Former@physicist.net

**Comment by tylerjereddy** [comment]
This looks fun! ping @person142 any chance this gets reviewed before we branch in 6 days?

**Comment by tylerjereddy** [comment]
@person142 What do you think? Do I need to bump the milestone again?

**Comment by person142** [comment]
Unfortunately yes.

**Comment by tylerjereddy** [comment]
@FormerPhysicist @person142 No review activity in the last 6 months--I'm going to have to bump the milestone soon given branching in ~1 week unless you tell me otherwise.

My Fortran skills are improving but probably not strong enough to review this with confidence yet.

**Comment by tylerjereddy** [comment]
@FormerPhysicist @person142 No activity here in about a year--I'll probably have to bump the milestone?

**Comment by h-vetinari** [comment]
Trying my hand at carrying this PR in #13310

**Comment by h-vetinari** [comment]
To avoid this PR slipping another release, could we have some (re-)review?

**Comment by h-vetinari** [comment]
Ping @person142 @tylerjereddy

**Comment by steppi** [comment]
@FormerPhysicist, @h-vetinari, @person142, @tylerjereddy

It's sad to see this languish so long. I'm a former mathematician and think I'm qualified to help review this. I can only spend 1-2 hours per week, but would it be helpful if I combed through and produced a technical report detailing the workings of the code that could aid a maintainer in their review process? At the rate of 1-2 hours per week I think I could have something in a couple months. I've been using some of my free time to brush up on my math anyway and had already begun diving back into complex analysis.

**Comment by tylerjereddy** [comment]
@steppi I suspect any help would be appreciated yeah. As Ralf noted, the combination of Fortran and very specialized math make this a big tricky for the current maintainer team.

**Comment by steppi** [comment]
As I start I can see if I'll have time this weekend or next to read Buhring's paper and write a clean and well documented reference implementation in pure python.

**Comment by tylerjereddy** [comment]
Bumping the milestone on this PR for the usual reasons--if activity does suddenly pick up here in the next few weeks feel free to ping me for milestone adjustment.

**Comment by steppi** [comment]
@FormerPhysicist, @h-vetinari, @person142, @tylerjereddy

I started looking into this. I think ideally it could be split into three pull requests to make review more manageable. One to address #8083, one to expand the regions where the (1-z) and z/(1-z) transformations are used, and one to address #8054. These are independent of each other and some are easier to review than others.

 Also I propose instead of using Buhring's analytic continuation series to use the series from [this paper](https://arxiv.org/abs/1306.2046) by José Luis López, and Nico M. Temme. It doesn't require special handling when `b - a` is an integer, making it simpler to understand in my opinion, and it also appears to have better convergence in general. Instead of implementing in Fortran, I think it could be implemented in Cython and dispatched in `_hypergeometric.pxd` similar to what's being done currently with `hyp1f1`. I imagine there would be a (slight?) performance impact but I think it's worth it in terms of the gain in ease of review and maintenance.

Thoughts?

**Comment by person142** [comment]
> Thoughts?

Generally anything involving less Fortran 77 code is good; that's why you're seeing this somewhat in-progress movement of hypergeometric functions into Cython.

**Comment by steppi** [comment]
@person142 Yeah, that makes a lot of sense. I'll take a crack at putting together a Cython implementation of the series from the paper I linked when I have free time over the weekends.

**Comment by person142** [comment]
I'll also note that any PRs moving the existing hypergeometric code into Cython (in small chunks at a time) will be gladly approved.

**Comment by steppi** [comment]
Cool. I'd be happy to chip away at that in my spare time.

**Comment by tylerjereddy** [comment]
I'll bump the milestone here again--my impression is that we're actually going to favor the effort by @steppi to avoid Fortran stuff anyway? So not sure if we can just drop the milestone entirely?

**Comment by person142** [comment]
Yes, I think this has been superseded at this point so I'm going to close. Please correct me if wrong.

**Pull request closed without merging**